Implement command palette widget

hanjinliu opened the pull request:
# Description

This PR implements a command palette widget.

![viewer-output](https://user-images.githubusercontent.com/40591297/212549098-17e42796-6114-4bb2-b699-3b61a61d32ea.gif)

Still work-in-progress, but here's the list of required functinalities and how much of them this PR has done (See #5430).

> Search logic: fuzzy matching ranked by match scoring

&rarr; fuzzy matching is not implemented yet. Just match by `__contains__` for now.

> Command context: breadcrumbs for each search result (e.g., segment_blobs_and_things with_membranes [plugin name] > gaussian blur or preferences > plugin settings > plugin API) 

&rarr; Contexts are auto-generated from the command ID (e.g. "napari:layer:XYZ" &rarr; "napari > layer > XYZ")

> Command coverage: napari menus & submenus, layer context menus, all plugin command/menu contributions as specified in the manifest

&rarr; Currently, only those commands registered in the `app-model` way are available (so some commands such as "Open Files ..." are not available).

> Action effects: there would be an indication in the UI that shows whether it will run the command, or direct to a dialog and additional input required from user

&rarr; Not yet, but current implementation is already prepared for it.

> Unavailable commands: gray out command options that aren’t available given the context. For example “split RGB” would be grayed out if the user hasn’t selected an image layer

&rarr; The "when" parameters of commands are evaluated and unavaliable ones are grayed out.


## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References


#5430

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

hanjinliu committed:
command palette installed correctly

hanjinliu committed:
command worked

hanjinliu committed:
generate command palette from app_model menus

hanjinliu committed:
fix for tests

DragaDoncila commented on napari/components/command_palette/_api.py:
```diff
@@ -0,0 +1,235 @@
+
+
+@inspect.signature
+def register_without_func(
```
When would we have a `func`less command? What would this be useful for?

DragaDoncila commented on napari/components/command_palette/_api.py:
```diff
@@ -0,0 +1,235 @@
+        title: str | None,
+        desc: str | None = None,
+        tooltip: str | None = None,
+        when: Callable[..., bool] = None,
```
I will need to double check this but for some reason I remember `app-model` commands having both `when` and `enablement` clauses and I feel like `when` is whether is shows up in the menu at all and `enablement` is whether it's greyed out or not. In which case we'd probably be wanting `enablement` here? Again I need to check on this though.

DragaDoncila commented on napari/components/command_palette/_api.py:
```diff
@@ -0,0 +1,235 @@
+        ...
+
+    @overload
+    def register(
```
Why is it that both `CommandPalette` and `CommandGroup` have the same overloaded `register` methods? Are you ever registered with one but not the other? Can the `CommandGroup` not just be an additional attribute of the `Command`?

DragaDoncila reviewed (commented):
This is awesome Hanjin, thanks for getting it off the ground!

I left a couple of general comments to aid my own understanding. Otherwise I have a few early concerns about the code:

- I notice we've added the command palette as a new `ViewerModel` argument and I suspect we probably don't want to do that, as I see the command palette as something purely internal - I don't imagine a user creating a viewer with a different command palette. But maybe that's just me 😆 In general I see the command palette as a widget layer on top of `app-model`'s commands rather than needing a new model
- I suspect we don't want the command palette to live in `components` if it has any GUI/Qt related code, which it looks like it does. I'm not sure where the best location would be but again, I would expect most of the code to be purely providing a GUI interface for `app-model`'s commands
- We may want the widget itself to be contributed to `napari/superqt`, or even depend on yours if you plan on maintaining it separately from `napari`?

I'm not the best at answering these kinds of code structure/where to put things questions. Maybe @Czaki has suggestions in that vein?

hanjinliu committed:
merge main

hanjinliu committed:
merging

hanjinliu committed:
sort commands

hanjinliu committed:
match score

hanjinliu committed:
match score

hanjinliu committed:
minor fixes

hanjinliu committed:
add test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni committed:
Merge branch 'main' into command-palette

hanjinliu committed:
don't use Command class

jni committed:
Don't add empty actions to the palette

jni committed:
directly instantiate palette rather than creating widget in model

jni committed:
Directly create QCommandPalette and add app_model actions

jni committed:
Remove unused command palette model

jni committed:
Remove accidental re-add of commands file

jni committed:
Connect app model changes to command palette list changes

jni committed:
Fix test

jni committed:
Add test checking that palette syncs when adding and removing commands

hanjinliu committed:
fix typing

hanjinliu committed:
fix typing

hanjinliu committed:
fix return annotation of flags()

lucyleeow commented on napari/_qt/widgets/qt_command_palette.py:
```diff
@@ -0,0 +1,431 @@
+        self.hide()
+
+        app = get_app_model()
+        # this appears to be a flat list of menu items, even though the
+        # type hint suggests menu or submenu
```
What do you mean 'flat list of menu items'? Is this not what `list[MenuOrSubmenu]` implies?

lucyleeow commented on napari/_qt/widgets/qt_command_palette.py:
```diff
@@ -0,0 +1,431 @@
+        app = get_app_model()
+        # this appears to be a flat list of menu items, even though the
+        # type hint suggests menu or submenu
+        menu_items = app.menus.get_menu(app.menus.COMMAND_PALETTE_ID)
+        self.extend_command(
```
Would we be able to just get all commands from the app-model command registry? If we ensure all actions we don't want to include have `palette=False`, this should be okay?

I think this would allow us to get rid of the complexity of `_on_app_menus_changed` and there should be a command registry changed event we could connect to.

Thinking about it more, we could also just update the command list just before the user opens the command palette, instead of updating whenever there has been a change to the command registry (/menu change as you have done here). Not clear on pro/cons, I guess updating just before opening the command palette may make it slower to open the command palette? 

This would also add any runtime user action additions (which I think we want to support with `bind_key` and/or `register_action`, see #7312), to the command palette. I think that would be nice.

(this would also be useful when we include plugin contributions, which we *may* want to be present more than once in the menu, once in the plugins menu and again as dictated by the menu contribution - I can't recall what we ultimately decided regarding this)

hanjinliu replied:
The reason why @jni and I decided to put a separate command list in command palette is because the command palette has different order of commands. In the current situation, the recently executed commands will appear on the top, but in the future "my favorite commands" or "frequently used commands" will also rely on this.

> Thinking about it more, we could also just update the command list just before the user opens the command palette, instead of updating whenever there has been a change to the command registry (/menu change as you have done here). 

I noticed this only recently 😅 totally agree.

lucyleeow commented on napari/_qt/widgets/qt_command_palette.py:
```diff
@@ -0,0 +1,431 @@
+        return False
+
+
+def _match_score(action: CommandRule, input_text: str) -> float:
```
Just a question, is it just for simplicity that we want to do this simple matching atm?

I would have thought `difflib` would have a function that does more complex matching, without too much effort on our part?

hanjinliu replied:
Yes, I think we should switch to fuzzy matching in the future, but for now, I cannot find any simple way to do that.

`difflib.SequenceMatcher` is not suitable for the intuitive text matching, even though I tried several ways to calculate matching scores. For example, when user type "op fi", it should match "napari > window > file > open file(s) ..." very well. However, what happens here is that `difflib.SequenceMatcher` picks "napari wind[o]w" as the first match, and the letters "napari > windo[w > file > o]pen file(s)" are considered "deletion". As the result, the matching score between "op fi" and "napari > window > file > open file(s) ..." is very low.

I think the local sequence alignment algorithm (which is famous in the field of bioinformatics) is useful for this purpose, but it will be very slow if we implement it in pure python. As far as I know, an efficient algorithm is implemented in `biopython` but I'm not sure if it accepts any string inputs.

lucyleeow replied:
Interesting! Thanks for the detailed response, I did not realise it was so complex. I totally agree this is something that can be looked into later.

lucyleeow commented on napari/_qt/widgets/qt_command_palette.py:
```diff
@@ -0,0 +1,431 @@
+        if command is None:
+            return
+        _exec_action(command)
+        # move to the top
```
```suggestion
        # move to the top, as we iteratively look for matches
```

lucyleeow commented on napari/_qt/widgets/qt_command_palette.py:
```diff
@@ -0,0 +1,431 @@
+        """Return true if the command can be executed."""
+        index = self._selected_index
+        command = self.command_at(index)
+        if command is None:
```
In what case would `command` be `None`?

hanjinliu replied:
`command_at` calls `indexWidget` inside, which may be None if widget does not exist at the given index. I *think* this happens only when user clicked on the empty region. As the return annotation of `indexWidget` is `QWidget | None`, we should consider None may be returned in some very specific situation.

lucyleeow commented on napari/_qt/widgets/qt_command_palette.py:
```diff
@@ -0,0 +1,431 @@
+    def iter_command(self) -> Iterator[CommandRule]:
+        """Iterate over all the commands registered to this command list widget."""
+        for i in range(self.model().rowCount()):
+            if not self.isRowHidden(i):
```
Just a question, how does `QListView` and the hidden row stuff work?

hanjinliu replied:
If more than`max_matches` commands matched, list will be filled with visible rows. If not, only the texts of first rows are updated and other rows are hidden. Does this answer your question?

lucyleeow replied:
Thanks. I guess I was wondering why we can't just set the number of rows to be the max of commands matched or 80, but I assume this is because I don't understand Qt or widgets :sweat_smile:

lucyleeow reviewed (commented):
Firstly, I think a command palette would be an amazing addition (and it works quite nicely), thanks for all your work.

Maybe we could add some more docstrings, this would help with review and code readability in future.

Context, I think there are currently 3 separate contexts (attached to viewer, qt main window and layerlist). AFAICT we can get each of these contexts with `napari._app_model.context.get_context`, so I don't think we need to keep a separate store of contexts for command palette.

I do appreciate that it is annoying that we have 3 separate contexts (there is some related discussion in #7101) and we've discussed that it would be ideal if we could have a single context, with parent/child hierarchical relationship (current app-model context has root and parent context functionality). There were difficulties in this though, so I don't foresee this being resolved quickly

This is an initial look over, before we look into it deeper with Draga and Juan at pycon au sprints

hanjinliu committed:
Merge branch 'napari:main' into command-palette

jni committed:
Merge branch 'main' into command-palette

jni committed:
Use None instead of QVariant() for empty data

PySide doesn't support QVariant and asks to instead use native Python
types. PyQt doesn't seem to mind so let's go with it.

See https://stackoverflow.com/questions/56117996/qvariant-alternative-when-migrating-from-pyqt5-to-pyside2

jni committed:
Explicitly parent line and command list widget

jni committed:
Revert "Explicitly parent line and command list widget"

This reverts commit 94ba7a2da8e32b3140e7ae9c88aa37fc6ccb4b2d.

hanjinliu committed:
update command palette font size and add placeholder text

hanjinliu committed:
remove update()

jni committed:
Merge branch 'main' into command-palette
